File: lib/features/horoscope/domain/services/horoscope_service.dart
import '../entities/horoscope_entity.dart';
import '../../../../config/constants/app_constants.dart';

class HoroscopeService {
  static final Map<String, Map<String, dynamic>> _signData = {
    'Aries': {
      'hindi': 'मेष',
      'luckyNumber': 9,
      'luckyColor': 'Red',
      'luckyGem': 'Red Coral',
      'compatible': ['Leo', 'Sagittarius', 'Gemini'],
      'predictions': [
        'Mars energizes you today. Seize new opportunities with courage.',
        'Your dynamic nature attracts attention. Lead with confidence.',
        'A challenging but rewarding day lies ahead. Stay determined.',
      ],
    },
    'Taurus': {
      'hindi': 'वृषभ',
      'luckyNumber': 6,
      'luckyColor': 'Green',
      'luckyGem': 'Diamond',
      'compatible': ['Virgo', 'Capricorn', 'Cancer'],
      'predictions': [
        'Venus blesses your relationships today. Show your appreciation.',
        'Financial opportunities emerge. Trust your practical instincts.',
        'Patience is your greatest virtue. Steady progress continues.',
      ],
    },
    'Gemini': {
      'hindi': 'मिथुन',
      'luckyNumber': 5,
      'luckyColor': 'Yellow',
      'luckyGem': 'Emerald',
      'compatible': ['Libra', 'Aquarius', 'Aries'],
      'predictions': [
        'Mercury sharpens your communication. Express your ideas boldly.',
        'Social connections bring unexpected benefits today.',
        'Your curious mind leads to fascinating discoveries.',
      ],
    },
    'Cancer': {
      'hindi': 'कर्क',
      'luckyNumber': 2,
      'luckyColor': 'White',
      'luckyGem': 'Pearl',
      'compatible': ['Scorpio', 'Pisces', 'Taurus'],
      'predictions': [
        'The Moon illuminates your emotional intelligence. Trust your intuition.',
        'Family matters require gentle attention. Your empathy heals.',
        'Home and security feel especially important. Nurture your foundations.',
      ],
    },
    'Leo': {
      'hindi': 'सिंह',
      'luckyNumber': 1,
      'luckyColor': 'Gold',
      'luckyGem': 'Ruby',
      'compatible': ['Aries', 'Sagittarius', 'Libra'],
      'predictions': [
        'The Sun empowers your natural leadership. Shine brightly today.',
        'Creative projects flourish under your enthusiastic touch.',
        'Recognition and praise come your way. Accept graciously.',
      ],
    },
    'Virgo': {
      'hindi': 'कन्या',
      'luckyNumber': 5,
      'luckyColor': 'Navy Blue',
      'luckyGem': 'Emerald',
      'compatible': ['Taurus', 'Capricorn', 'Cancer'],
      'predictions': [
        'Mercury aids your analytical abilities. Details matter today.',
        'Health and wellness focus brings positive results.',
        'Practical solutions emerge from careful observation.',
      ],
    },
    'Libra': {
      'hindi': 'तुला',
      'luckyNumber': 6,
      'luckyColor': 'Pink',
      'luckyGem': 'Diamond',
      'compatible': ['Gemini', 'Aquarius', 'Leo'],
      'predictions': [
        'Venus harmonizes your relationships. Balance is your strength.',
        'Diplomatic skills resolve long-standing conflicts gracefully.',
        'Beauty and aesthetics bring joy. Indulge your artistic senses.',
      ],
    },
    'Scorpio': {
      'hindi': 'वृश्चिक',
      'luckyNumber': 9,
      'luckyColor': 'Deep Red',
      'luckyGem': 'Red Coral',
      'compatible': ['Cancer', 'Pisces', 'Capricorn'],
      'predictions': [
        'Mars intensifies your focus. Transformation is underway.',
        'Hidden truths surface, bringing clarity to complex situations.',
        'Your magnetic presence draws powerful allies to your cause.',
      ],
    },
    'Sagittarius': {
      'hindi': 'धनु',
      'luckyNumber': 3,
      'luckyColor': 'Purple',
      'luckyGem': 'Yellow Sapphire',
      'compatible': ['Aries', 'Leo', 'Aquarius'],
      'predictions': [
        'Jupiter expands your horizons. Adventure beckons you today.',
        'Philosophical insights lead to profound understanding.',
        'Travel or higher learning brings exciting opportunities.',
      ],
    },
    'Capricorn': {
      'hindi': 'मकर',
      'luckyNumber': 8,
      'luckyColor': 'Black',
      'luckyGem': 'Blue Sapphire',
      'compatible': ['Taurus', 'Virgo', 'Scorpio'],
      'predictions': [
        'Saturn rewards your disciplined efforts with lasting success.',
        'Career advancement is highlighted. Showcase your expertise.',
        'Long-term planning pays off. Your patience bears fruit.',
      ],
    },
    'Aquarius': {
      'hindi': 'कुम्भ',
      'luckyNumber': 4,
      'luckyColor': 'Blue',
      'luckyGem': 'Blue Sapphire',
      'compatible': ['Gemini', 'Libra', 'Sagittarius'],
      'predictions': [
        'Saturn and Rahu bring innovative thinking. Embrace change.',
        'Social causes and humanitarian efforts gain momentum.',
        'Original ideas set you apart. Think outside conventions.',
      ],
    },
    'Pisces': {
      'hindi': 'मीन',
      'luckyNumber': 3,
      'luckyColor': 'Sea Green',
      'luckyGem': 'Yellow Sapphire',
      'compatible': ['Cancer', 'Scorpio', 'Capricorn'],
      'predictions': [
        'Jupiter and Neptune enhance your spiritual sensitivity.',
        'Creative and artistic pursuits bring deep fulfillment.',
        'Compassion and healing abilities are at their peak today.',
      ],
    },
  };

  static final Map<String, List<String>> _lovePredictions = {
    'Daily': [
      'Romance is in the air. Express your feelings openly.',
      'A special connection deepens today. Cherish every moment.',
      'Misunderstandings clear up with honest communication.',
      'Your partner appreciates your thoughtful gestures today.',
      'New romantic possibilities emerge unexpectedly.',
    ],
    'Weekly': [
      'This week brings emotional clarity in relationships.',
      'Open your heart to new possibilities in love.',
      'Communication with your partner improves significantly.',
      'A week of romantic moments and deeper bonds.',
      'Trust builds stronger foundations in your relationship.',
    ],
  };

  static final Map<String, List<String>> _careerPredictions = {
    'Daily': [
      'Professional recognition comes from consistent effort.',
      'A senior colleague offers valuable guidance today.',
      'New project opportunities align with your skills.',
      'Decision-making abilities are sharp. Act decisively.',
      'Team collaboration leads to breakthrough results.',
    ],
    'Weekly': [
      'Career advancement is on the horizon this week.',
      'Financial gains through professional expertise.',
      'Leadership qualities shine in team settings.',
      'An important business meeting yields positive outcomes.',
      'New career opportunities deserve serious consideration.',
    ],
  };

  static HoroscopeEntity getHoroscope(
    String zodiacSign,
    String period,
    DateTime date,
  ) {
    final data = _signData[zodiacSign] ?? _signData['Aries']!;
    final predictions = data['predictions'] as List<String>;
    final dayIndex = date.day % predictions.length;

    // Generate varied predictions based on date
    final seed = date.day + date.month * 31 + AppConstants.zodiacSigns.indexOf(zodiacSign);

    final lovePreds = _lovePredictions['Daily']!;
    final careerPreds = _careerPredictions['Daily']!;

    final mainPrediction = _buildPrediction(zodiacSign, period, date, predictions[dayIndex]);

    return HoroscopeEntity(
      zodiacSign: zodiacSign,
      zodiacSignHindi: data['hindi'] as String,
      date: date,
      prediction: mainPrediction,
      love: lovePreds[seed % lovePreds.length],
      career: careerPreds[(seed + 2) % careerPreds.length],
      health: _buildHealthPrediction(seed),
      finance: _buildFinancePrediction(seed),
      luckyNumber: data['luckyNumber'] as int,
      luckyColor: data['luckyColor'] as String,
      luckyGem: data['luckyGem'] as String,
      compatibleSigns: List<String>.from(data['compatible'] as List),
      rating: (seed % 3) + 3, // 3-5 stars
    );
  }

  static String _buildPrediction(
    String sign,
    String period,
    DateTime date,
    String base,
  ) {
    final periodText = switch (period) {
      'Daily' => 'Today',
      'Weekly' => 'This week',
      'Monthly' => 'This month',
      'Yearly' => 'This year',
      _ => 'Today',
    };

    final extras = [
      '$periodText, the cosmic energies favor bold decisions. $base',
      'The stars align in your favor. $base Your efforts will be rewarded.',
      '$base Maintain focus and stay grounded as opportunities unfold.',
      'Planetary movements bring positive changes. $base Trust the cosmic plan.',
      '$base $periodText brings clarity and renewed purpose to your journey.',
    ];

    final idx = (date.day + AppConstants.zodiacSigns.indexOf(sign)) % extras.length;
    return extras[idx];
  }

  static String _buildHealthPrediction(int seed) {
    const preds = [
      'Energy levels are high. Channel them into physical activity.',
      'Rest and relaxation are essential today. Listen to your body.',
      'A good day for starting a health routine. Small steps matter.',
      'Mental wellness deserves attention. Practice mindfulness.',
      'Avoid overexertion. Pace yourself through the day.',
    ];
    return preds[seed % preds.length];
  }

  static String _buildFinancePrediction(int seed) {
    const preds = [
      'Avoid impulsive spending. Save for long-term goals.',
      'A favorable day for investments and financial planning.',
      'Unexpected income is possible. Manage it wisely.',
      'Review your budget and cut unnecessary expenses.',
      'Business ventures initiated today show promising returns.',
    ];
    return preds[(seed + 3) % preds.length];
  }

  static List<String> getAllZodiacSigns() {
    return AppConstants.zodiacSigns;
  }
}

File: app/src/main/java/com/vaibhav/kundli/domain/service/HoroscopeService.kt
package com.vaibhav.kundli.domain.service

import com.vaibhav.kundli.domain.model.HoroscopeData
import com.vaibhav.kundli.util.Constants
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class HoroscopeService @Inject constructor() {

    private val luckyNumbers = listOf(9, 6, 5, 2, 1, 5, 6, 9, 3, 8, 4, 7)
    private val luckyColors = listOf("Red", "Green", "Yellow", "White", "Gold", "Navy Blue",
        "Pink", "Maroon", "Orange", "Black", "Electric Blue", "Sea Green")
    private val luckyGems = listOf("Ruby", "Emerald", "Agate", "Pearl", "Ruby", "Emerald",
        "Diamond", "Red Coral", "Yellow Sapphire", "Blue Sapphire", "Amethyst", "Pearl")

    private val compatibleSigns = listOf(
        listOf("Leo", "Sagittarius", "Gemini"),
        listOf("Virgo", "Capricorn", "Cancer"),
        listOf("Libra", "Aquarius", "Aries"),
        listOf("Scorpio", "Pisces", "Taurus"),
        listOf("Aries", "Sagittarius", "Gemini"),
        listOf("Taurus", "Capricorn", "Cancer"),
        listOf("Gemini", "Aquarius", "Leo"),
        listOf("Cancer", "Pisces", "Virgo"),
        listOf("Leo", "Aries", "Aquarius"),
        listOf("Virgo", "Taurus", "Scorpio"),
        listOf("Libra", "Gemini", "Sagittarius"),
        listOf("Scorpio", "Cancer", "Capricorn")
    )

    private val dailyPredictions = mapOf(
        "Aries" to listOf(
            "A new opportunity presents itself today. Seize it with confidence and enthusiasm. Your natural leadership will guide you well.",
            "Financial matters require careful attention today. Avoid impulsive spending and focus on long-term goals.",
            "Your energy levels are high today. Channel this into productive activities and avoid unnecessary conflicts.",
            "Travel or communication brings good news. Stay open to unexpected connections that could benefit your career.",
            "Focus on self-care and wellness today. Rest and reflection will lead to important insights about your life path."
        ),
        "Taurus" to listOf(
            "Stability is your strength today. Financial investments made now will yield good returns in the future.",
            "Relationships need nurturing. Spend quality time with loved ones and express your feelings openly.",
            "A creative project demands your attention. Your practical approach will turn visions into reality.",
            "Health and diet need focus. A small change in your routine will lead to significant improvements.",
            "Trust your instincts in professional matters. Your judgment is sound and colleagues respect your opinion."
        ),
        "Gemini" to listOf(
            "Communication is your superpower today. Use it wisely in both personal and professional settings.",
            "Mental agility helps you solve a complex problem. Share your ideas freely; they'll be well-received.",
            "A social gathering brings unexpected opportunities. Networking opens new doors for career advancement.",
            "Dual energies may cause indecision. Make a firm choice and stick with it for best results.",
            "Creative inspiration strikes strongly today. Document your ideas before they slip away."
        ),
        "Cancer" to listOf(
            "Home and family matters take center stage. Emotional bonds are strengthened through honest conversation.",
            "Intuition guides you correctly today. Trust those gut feelings, especially in financial decisions.",
            "A nurturing gesture towards a friend or family member creates lasting positive karma in your life.",
            "Career opportunities connected to your caring nature emerge. Consider roles where you help others.",
            "Emotions run deep today. Allow yourself to feel without judgment, but maintain perspective."
        ),
        "Leo" to listOf(
            "Your charisma attracts positive attention today. Step into the spotlight with confidence and grace.",
            "Creative endeavors receive recognition. Share your talents generously and watch your reputation grow.",
            "Leadership is called for in a group situation. Your natural authority inspires others to follow.",
            "Romance is highlighted today. Express your feelings boldly and generously for best results.",
            "Financial generosity pays dividends. Give back to those who have supported your journey."
        ),
        "Virgo" to listOf(
            "Attention to detail serves you well today. A thorough approach to work earns appreciation from superiors.",
            "Health routines established today will have lasting benefits. Small consistent efforts compound over time.",
            "Analytical skills help solve a problem others have struggled with. Your expertise is invaluable.",
            "Service to others brings deep satisfaction. Volunteering or helping a friend lifts your own spirits.",
            "Organization and planning for upcoming events reduces stress. Your methodical approach ensures success."
        ),
        "Libra" to listOf(
            "Balance and harmony in relationships are your focus today. Diplomacy resolves a long-standing conflict.",
            "Aesthetic sensibilities are heightened. Artistic pursuits bring joy and creative breakthroughs.",
            "Partnerships, both personal and professional, are favored. Collaboration yields better results than solo efforts.",
            "Justice matters occupy your thoughts. Standing up for fairness earns respect from peers.",
            "Social grace and charm open doors. An important introduction at a social event changes your trajectory."
        ),
        "Scorpio" to listOf(
            "Deep transformation is underway. Embrace change rather than resisting it for personal growth.",
            "Research and investigation bring hidden truths to light. Knowledge gained now proves valuable later.",
            "Intensity in relationships can be channeled productively. Passion directed into creative work yields masterpieces.",
            "Financial acumen is sharp today. Complex investment strategies or negotiations favor you.",
            "Spiritual insights arrive through meditation or dreams. Pay attention to symbolic messages from your subconscious."
        ),
        "Sagittarius" to listOf(
            "Adventure and exploration call to your spirit. Even a mental journey through books broadens your horizons.",
            "Philosophical discussions and higher learning bring fulfillment. Share your wisdom with those who seek it.",
            "Optimism is your greatest asset today. Your positive outlook inspires and motivates those around you.",
            "Travel plans or educational pursuits move forward favorably. Take the next step with enthusiasm.",
            "Freedom and independence are essential to your wellbeing today. Create space for spontaneity in your schedule."
        ),
        "Capricorn" to listOf(
            "Ambition and determination drive you toward important goals. Hard work today lays the foundation for future success.",
            "Professional achievements receive recognition. Your reputation for reliability opens new career opportunities.",
            "Practical wisdom guides financial decisions. Conservative approaches to money management prove wise.",
            "Structure and discipline in your daily routine yield impressive results. Consistency is your key to success.",
            "Long-term planning and goal setting bring clarity. Map out your path to success with concrete steps."
        ),
        "Aquarius" to listOf(
            "Innovation and original thinking set you apart today. Your unconventional approach solves problems others cannot.",
            "Humanitarian concerns motivate meaningful action. Contributing to community welfare brings deep satisfaction.",
            "Technology and digital domains offer opportunities. Stay updated with the latest developments in your field.",
            "Friendships and social networks prove valuable. A connection in your circle provides crucial assistance.",
            "Independence and intellectual freedom fuel your creativity. Trust your unique vision and share it boldly."
        ),
        "Pisces" to listOf(
            "Intuition and spiritual insight are heightened today. Trust the subtle messages that come through your inner voice.",
            "Creative and artistic endeavors flow naturally. Allow imagination to guide your work without overthinking.",
            "Compassion for others opens your heart and theirs. A simple act of kindness creates ripples of positive change.",
            "Dreams may carry important messages. Keep a journal nearby to capture insights from your subconscious mind.",
            "Spiritual practices and meditation bring peace and clarity. Quiet time alone helps you reconnect with your purpose."
        )
    )

    private val lovePredictions = listOf(
        "Venus blesses your relationships. Express love openly and watch bonds deepen.",
        "Communication in relationships is key today. Listen as much as you speak.",
        "A romantic surprise brings joy. Be open to spontaneous moments of connection.",
        "Patience in love matters. Allow things to unfold naturally at their own pace.",
        "New romantic possibilities emerge. Stay open to meeting someone special.",
        "Existing relationships are strengthened through shared experiences today.",
        "Misunderstandings can be cleared with honest conversation. Speak your truth kindly.",
        "Love requires compromise today. Find the middle ground for harmony.",
        "Your charm is magnetic. Social settings bring wonderful romantic possibilities.",
        "Deep emotional connection with a partner brings profound satisfaction today."
    )

    private val careerPredictions = listOf(
        "Career advancements are indicated. Take initiative on important projects.",
        "New professional opportunities present themselves. Evaluate carefully before committing.",
        "Teamwork and collaboration lead to exceptional results. Value others' contributions.",
        "Your expertise is recognized. Share your knowledge generously with colleagues.",
        "Financial gains through career efforts. Negotiate confidently for what you deserve.",
        "Creative thinking solves a workplace challenge. Your innovative approach impresses management.",
        "Networking events bring valuable professional connections. Attend social gatherings.",
        "Focus on skill development. Learning something new enhances your career prospects.",
        "Leadership opportunities arise. Step forward with confidence and clear communication.",
        "Work-life balance needs attention. Setting boundaries improves overall productivity."
    )

    private val healthPredictions = listOf(
        "Physical vitality is strong. Channel this energy into exercise and healthy activities.",
        "Mental health needs attention. Mindfulness practices bring peace and clarity.",
        "Digestive health benefits from lighter, more frequent meals today.",
        "Rest and recovery are important. Don't push yourself beyond reasonable limits.",
        "Outdoor activity brings both physical and mental benefits. Spend time in nature.",
        "Stress management techniques are especially helpful today. Practice deep breathing.",
        "Nutrition choices significantly impact your energy levels. Choose foods wisely.",
        "Joint health and flexibility improve with gentle stretching. Make time for yoga.",
        "Your immune system is strong. Continue healthy habits that support your wellbeing.",
        "Eye strain from screens is possible. Take regular breaks and rest your eyes."
    )

    private val financePredictions = listOf(
        "Financial opportunities knock. Be ready to act decisively on good investments.",
        "Budget review leads to discovering unnecessary expenses. Trim where possible.",
        "Income from an unexpected source brightens your financial picture today.",
        "Long-term financial planning benefits from expert consultation. Seek advice.",
        "Avoid risky financial ventures today. Stick to proven strategies.",
        "Generosity brings its own rewards. Charitable giving creates positive financial karma.",
        "Career and financial matters are linked. Professional advancement improves income.",
        "Real estate or property matters look favorable for investment consideration.",
        "Review your savings strategy. Small adjustments now create significant long-term gains.",
        "Financial discipline is your strength. Avoid impulsive purchases and stay on budget."
    )

    fun getHoroscope(zodiacSign: String, period: String = "daily", date: Date = Date()): HoroscopeData {
        val signIndex = Constants.ZODIAC_SIGNS.indexOf(zodiacSign).takeIf { it >= 0 } ?: 0
        val calendar = Calendar.getInstance()
        calendar.time = date
        val seed = calendar.get(Calendar.DAY_OF_YEAR) + signIndex
        val rand = Random(seed.toLong())

        val predictions = dailyPredictions[zodiacSign] ?: dailyPredictions["Aries"]!!
        val predictionText = predictions[rand.nextInt(predictions.size)]
        val love = lovePredictions[rand.nextInt(lovePredictions.size)]
        val career = careerPredictions[rand.nextInt(careerPredictions.size)]
        val health = healthPredictions[rand.nextInt(healthPredictions.size)]
        val finance = financePredictions[rand.nextInt(financePredictions.size)]
        val rating = 3 + rand.nextInt(3) // 3-5 stars

        return HoroscopeData(
            zodiacSign = zodiacSign,
            zodiacSignHindi = Constants.ZODIAC_SIGNS_HINDI.getOrElse(signIndex) { zodiacSign },
            date = date,
            prediction = predictionText,
            love = love,
            career = career,
            health = health,
            finance = finance,
            luckyNumber = luckyNumbers[signIndex],
            luckyColor = luckyColors[signIndex],
            luckyGem = luckyGems[signIndex],
            compatibleSigns = compatibleSigns[signIndex],
            rating = rating
        )
    }

    fun getAllSignsHoroscope(date: Date = Date()): List<HoroscopeData> {
        return Constants.ZODIAC_SIGNS.map { getHoroscope(it, "daily", date) }
    }
}
